packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/SyncConfigurationImpl.kt

```diff
@@ -303,3 +208,247 @@ internal class SyncConfigurationImpl(
     override val syncMode: SyncMode =
         if (partitionValue == null) SyncMode.FLEXIBLE else SyncMode.PARTITION_BASED
 }
+
+private interface ClientResetStrategyHelper {
+    fun initialize(nativeSyncConfig: RealmSyncConfigurationPointer)
+    fun onSyncError(session: SyncSession, appPointer: RealmAppPointer, error: SyncError)
+}
+
+private abstract class OnBeforeOnAfterHelper<T : SyncClientResetStrategy> constructor(
+    val strategy: T,
+    val configuration: InternalConfiguration
+) : ClientResetStrategyHelper {
+
+    abstract fun getResyncMode(): SyncSessionResyncMode
+    abstract fun getBefore(): SyncBeforeClientResetHandler
+    abstract fun getAfter(): SyncAfterClientResetHandler
+
+    override fun initialize(nativeSyncConfig: RealmSyncConfigurationPointer) {
+        RealmInterop.realm_sync_config_set_resync_mode(nativeSyncConfig, getResyncMode())
+        RealmInterop.realm_sync_config_set_before_client_reset_handler(
+            nativeSyncConfig,
+            getBefore()
+        )
+        RealmInterop.realm_sync_config_set_after_client_reset_handler(
+            nativeSyncConfig,
+            getAfter()
+        )
+    }
+}
+
+private class RecoverOrDiscardUnsyncedChangesHelper constructor(
+    strategy: RecoverOrDiscardUnsyncedChangesStrategy,
+    configuration: InternalConfiguration
+) : OnBeforeOnAfterHelper<RecoverOrDiscardUnsyncedChangesStrategy>(strategy, configuration) {
+
+    override fun getResyncMode(): SyncSessionResyncMode =
+        SyncSessionResyncMode.RLM_SYNC_SESSION_RESYNC_MODE_RECOVER_OR_DISCARD
+
+    override fun getBefore(): SyncBeforeClientResetHandler =
+        object : SyncBeforeClientResetHandler {
+            override fun onBeforeReset(realmBefore: FrozenRealmPointer) {
+                strategy.onBeforeReset(TypedFrozenRealmImpl(realmBefore, configuration))
+            }
+        }
+
+    override fun getAfter(): SyncAfterClientResetHandler =
```

Comment:
There are multiple implementations for `getAfter` and `getBefore` that really similar between them, is there any way to reuse the code?